File: tools/server_observer/src/static_map_cache.cpp
#include "static_map_cache.hpp"
#include "db_client.hpp"
#include <fstream>
#include <filesystem>
#include <iostream>
#include <vector>
#include <zstd.h>

namespace fs = std::filesystem;

StaticMapCache::StaticMapCache(const std::string& cache_dir)
    : cache_dir_(cache_dir)
    , s3_enabled_(false)
{
    fs::create_directories(cache_dir_);
    
    // Load existing cached map IDs
    for (const auto& entry : fs::directory_iterator(cache_dir_)) {
        if (entry.is_regular_file()) {
            std::string filename = entry.path().stem().string();
            if (filename.find("map_") == 0) {
                std::string map_id = filename.substr(4);
                saved_ids_.insert(map_id);
            }
        }
    }
}

StaticMapCache::StaticMapCache(const std::string& cache_dir, const S3Config& s3_config)
    : cache_dir_(cache_dir)
    , s3_enabled_(true)
    , db_client_(nullptr)
{
    fs::create_directories(cache_dir_);
    
    // Initialize S3 client
    s3_client_ = std::make_shared<S3Client>(s3_config);
    
    // Ensure bucket exists
    ensure_bucket_exists();
    
    // Load existing cached map IDs
    for (const auto& entry : fs::directory_iterator(cache_dir_)) {
        if (entry.is_regular_file()) {
            std::string filename = entry.path().stem().string();
            if (filename.find("map_") == 0) {
                std::string map_id = filename.substr(4);
                saved_ids_.insert(map_id);
            }
        }
    }
}

StaticMapCache::StaticMapCache(const std::string& cache_dir, const S3Config& s3_config, std::shared_ptr<DbClient> db_client)
    : cache_dir_(cache_dir)
    , s3_enabled_(true)
    , db_client_(db_client)
{
    fs::create_directories(cache_dir_);
    
    // Initialize S3 client
    s3_client_ = std::make_shared<S3Client>(s3_config);
    
    // Ensure bucket exists
    ensure_bucket_exists();
    
    // Load existing cached map IDs
    for (const auto& entry : fs::directory_iterator(cache_dir_)) {
        if (entry.is_regular_file()) {
            std::string filename = entry.path().stem().string();
            if (filename.find("map_") == 0) {
                std::string map_id = filename.substr(4);
                saved_ids_.insert(map_id);
            }
        }
    }
}

StaticMapCache::~StaticMapCache() {
    // Destructor
}

bool StaticMapCache::is_cached(const std::string& map_id) const {
    std::lock_guard<std::mutex> lock(lock_);
    return saved_ids_.find(map_id) != saved_ids_.end();
}

std::string StaticMapCache::save(const std::string& map_id, const json& static_map_data) {
    std::lock_guard<std::mutex> lock(lock_);
    
    std::string filename = "map_" + map_id + ".bin";
    fs::path path = fs::path(cache_dir_) / filename;
    
    // Check if already cached
    if (saved_ids_.find(map_id) != saved_ids_.end() && fs::exists(path)) {
        return path.string();
    }
    
    try {
        // Serialize JSON to string
        std::string json_str = static_map_data.dump();
        
        // Compress using zstd
        size_t compressed_size = ZSTD_compressBound(json_str.size());
        std::vector<uint8_t> compressed(compressed_size);
        
        size_t actual_size = ZSTD_compress(
            compressed.data(), compressed_size,
            json_str.data(), json_str.size(),
            3  // compression level
        );
        
        if (ZSTD_isError(actual_size)) {
            std::cerr << "Compression failed: " << ZSTD_getErrorName(actual_size) << std::endl;
            return "";
        }
        
        compressed.resize(actual_size);
        
        // Write to file
        std::ofstream file(path, std::ios::binary);
        if (!file.is_open()) {
            std::cerr << "Failed to open file for writing: " << path << std::endl;
            return "";
        }
        
        file.write(reinterpret_cast<const char*>(compressed.data()), compressed.size());
        file.close();
        
        saved_ids_.insert(map_id);
        std::cout << "Cached static map data for map_id " << map_id << " at " << path << std::endl;
        
        // Upload to S3 if enabled
        std::string s3_key;
        if (s3_enabled_) {
            s3_key = "static_maps/map_" + map_id + ".bin";
            if (upload_to_s3(path.string(), map_id)) {
                std::cout << "Uploaded static map data for map_id " << map_id << " to S3" << std::endl;
            } else {
                std::cerr << "Failed to upload static map data for map_id " << map_id << " to S3" << std::endl;
            }
        }
        
        // Record in database if enabled
        if (db_client_ && db_client_->is_connected()) {
            if (!db_client_->map_exists(map_id)) {
                // Extract version from static_map_data if available
                std::optional<std::string> version = std::nullopt;
                if (static_map_data.contains("version") && !static_map_data["version"].is_null()) {
                    version = static_map_data["version"].get<std::string>();
                }
                
                if (db_client_->insert_map(map_id, s3_key, version)) {
                    std::cout << "Recorded static map " << map_id << " in database" << std::endl;
                } else {
                    std::cerr << "Failed to record static map " << map_id << " in database" << std::endl;
                }
            }
        }
        
        return path.string();
    } catch (const std::exception& e) {
        std::cerr << "Failed to cache static map data for map_id " << map_id << ": " << e.what() << std::endl;
        return "";
    }
}

bool StaticMapCache::upload_to_s3(const std::string& local_path, const std::string& map_id) {
    if (!s3_client_) {
        std::cerr << "S3 client not initialized" << std::endl;
        return false;
    }
    
    try {
        std::string s3_key = "static_maps/map_" + map_id + ".bin";
        return s3_client_->upload_file(local_path, s3_key);
    } catch (const std::exception& e) {
        std::cerr << "Exception during S3 upload: " << e.what() << std::endl;
        return false;
    }
}

void StaticMapCache::ensure_bucket_exists() {
    if (!s3_client_) {
        return;
    }
    
    try {
        if (!s3_client_->bucket_exists()) {
            std::cout << "Creating S3 bucket..." << std::endl;
            if (!s3_client_->create_bucket()) {
                std::cerr << "Failed to create S3 bucket" << std::endl;
            }
        }
    } catch (const std::exception& e) {
        std::cerr << "Exception checking/creating S3 bucket: " << e.what() << std::endl;
    }
}

File: services/server_observer/src/static_map_cache.rs
use crate::db::DbClient;
use crate::s3_client::S3Client;
use serde_json::Value as JsonValue;
use std::collections::HashSet;
use std::sync::{Arc, Mutex};
use thiserror::Error;
use zstd::stream::encode_all;

#[derive(Debug, Error)]
pub enum StaticMapError {
    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),
    #[error("DB error: {0}")]
    Db(#[from] crate::db::DbClientError),
    #[error("S3 error: {0}")]
    S3(#[from] crate::s3_client::S3Error),
    #[error("Compression error: {0}")]
    Compression(String),
}

#[derive(Clone)]
pub struct StaticMapCache {
    saved_ids: Arc<Mutex<HashSet<String>>>,
    s3: S3Client,
    db: DbClient,
}

impl StaticMapCache {
    pub async fn new(s3: S3Client, db: DbClient) -> Result<Self, StaticMapError> {
        let mut ids = HashSet::new();
        // Pre-populate the in-memory cache from existing DB entries so
        // we don't re-upload or re-insert maps that are already known.
        for map_id in db.get_all_map_ids().await? {
            ids.insert(map_id);
        }
        Ok(Self {
            saved_ids: Arc::new(Mutex::new(ids)),
            s3,
            db,
        })
    }

    pub fn is_cached(&self, map_id: &str) -> bool {
        let ids = self.saved_ids.lock().unwrap();
        ids.contains(map_id)
    }

    pub async fn save(
        &self,
        map_id: &str,
        static_map_data: &JsonValue,
    ) -> Result<(), StaticMapError> {
        {
            let ids = self.saved_ids.lock().unwrap();
            if ids.contains(map_id) {
                return Ok(());
            }
        }

        let json_str = static_map_data.to_string();
        let compressed = encode_all(json_str.as_bytes(), 3)
            .map_err(|e| StaticMapError::Compression(e.to_string()))?;

        let s3_key = format!("static_maps/map_{}.bin", map_id);
        self.s3.upload_bytes(compressed, &s3_key).await?;

        if !self.db.map_exists(map_id).await? {
            let version = static_map_data
                .get("version")
                .and_then(|v| v.as_str())
                .map(|s| s.to_string());
            self.db
                .insert_map(map_id, &s3_key, version.as_deref())
                .await?;
        }

        let mut ids = self.saved_ids.lock().unwrap();
        ids.insert(map_id.to_string());

        Ok(())
    }
}

